lib/authentication.js

```diff
@@ -0,0 +1,172 @@
+    hints, // Array, [{ required: Array, optional: Array }, ...]
+    bestPercent // Number, percent of best test group
+  ) {
+    this.hints = hints;
+    this.bestPercent = bestPercent;
+  }
+  get valid() {
+    return !!this.bestPercent;
+  }
+  get strength() {
+    if (this.bestPercent < 25) {
+      return 'Very weak';
+    } else if (this.bestPercent < 40) {
+      return 'Weak';
+    } else if (this.bestPercent < 60) {
+      return 'Good';
+    } else if (this.bestPercent < 80) {
+      return 'Strong';
+    } else {
+      return 'Very strong';
+    }
+  }
+}
+
+const makeTest = (
+  // Function that checks the arguments on a test suite
+  tests, // Array, array of password/login tests
+  testGroups, // Array, [{ required: [testNames], optional: [testNames] }]
+  ...testArgs // Array, [password]/[login]/[login, password]
+  // Returns: AuthenticationStrength instance
+) => {
+  const test = testsNames => {
+    const testsHints = [];
+    testsNames.forEach(testName => {
+      const [test, hint] = tests[testName];
+      if (test(...testArgs)) testsHints.push(hint);
+    });
+    return testsHints;
+  };
+
+  const hints = [];
+  let bestPercent = 0;
+
+  testGroups.forEach(({ required, optional }) => {
+    const requiredHints = test(required);
+    const optionalHints = test(optional);
+
+    if (!requiredHints.length) {
+      const optTestsCount = optional.length;
+      const optHintsCount = optionalHints.length;
+      const percent = 100 * (optTestsCount - optHintsCount) / optTestsCount;
+      bestPercent = Math.max(bestPercent, percent);
+    }
+
+    hints.push({ required: requiredHints, optional: optionalHints });
+  });
+
+  hints.sort((prevHints, hints) => {
+    if (prevHints.required.length < hints.required.length) {
+      return -1;
+    } else if (prevHints.required.length === hints.required.length) {
+      if (prevHints.optional.length <= hints.optional.length) return -1;
+      else return 1;
+    } else {
+      return 1;
+    }
+  });
+
+  return new AuthenticationStrength(hints, bestPercent);
+};
+
+const checkLogin = (
+  // Function that tests the login
+  login, // String, login for testing
+  testGroups // Array, [{ required: [testNames], optional: [testNames] }]
+  // Returns: AuthenticationStrength instance
+) => (
+  makeTest(loginTests, testGroups, login)
+);
+
+const checkPassword = (
+  // Function that tests the password
+  password, // String, password for testing
+  testGroups // Array, [{ required: [testNames], optional: [testNames] }]
+  // Returns: AuthenticationStrength instance
+) => (
+  makeTest(passwordTests, testGroups, password)
+);
+
+const checkLoginPassword = (
+  // Function that tests the login with password
+  login, // String, login for testing
+  password, // String, password for testing
+  testsGroups // Array, [{ required: [testNames], optional: [testNames] }]
+  // Returns: AuthenticationStrength instance
+) => (
+  makeTest(loginPasswordTests, testsGroups, login, password)
```

Comment:
What's the use of passing `loginPasswordTests`, `loginTests`, `passwordTests` everywhere as they are global in this file and always the same?

Reply:
You can find out if your data passes the selected tests or not. I think I did not quite understand this question.

Reply:
Nevermind, I misunderstood the `makeTest`.